File: Client/src/_services/guild/GamersListView.ts
import { GamerStatus } from "../gamer/GamerStatus";
import { GamerRank } from "../gamer/GamerRank";
import { PenaltyStatus } from "./PenaltyStatus";
import { LoanStatus } from "./LoanStatus";
import { Dictionary } from "../../core";

export interface IGamersListView {
    id: string;
    characters: Array<string>;
    balance: number;
    penalties: Dictionary<IPenaltyView>;
    loans: Dictionary<ILoanView>;
    rank: GamerRank;
    status: GamerStatus;
    dateOfBirth: Date;

    GetLoan(id: string): ILoanView;
    GetPenalty(penaltyId: string): IPenaltyView;
}

export interface IPenaltyView {
    id: string;
    date: Date;
    amount: number;
    description: string;
    penaltyStatus: PenaltyStatus
}

export interface ILoanView {
    id: string;
    date: Date;
    expiredDate: Date;
    amount: number;
    description: string;
    loanStatus: LoanStatus
}

export class GamersListView implements IGamersListView {
    id: string;
    characters: Array<string>;
    balance: number;
    penalties: Dictionary<IPenaltyView>;
    loans: Dictionary<ILoanView>;
    rank: GamerRank;
    status: GamerStatus;
    dateOfBirth: Date;

    GetPenalty(penaltyId: string): IPenaltyView {
        if (this.penalties[penaltyId])
            return this.penalties[penaltyId]
        return {} as IPenaltyView;
    }

    GetLoan(id: string): ILoanView {
        if (this.loans[id])
            return this.loans[id]
        return {} as ILoanView;
    }

    constructor(id: string,
        characters: Array<string>,
        balance: number,
        penalties: Array<IPenaltyView>,
        loans: Array<ILoanView>,
        rank: GamerRank,
        status: GamerStatus,
        dateOfBirth: string
    ) {
        this.id = id;
        this.characters = characters;
        this.balance = balance;
        this.penalties = {};
        for (let i = 0; i < penalties.length; i++) {
            this.penalties[penalties[i].id] = penalties[i];
        }
        this.loans = {};
        for (let i = 0; i < loans.length; i++) {
            this.loans[loans[i].id] = loans[i];
        }
        this.rank = rank;
        this.status = status;
        this.dateOfBirth = new Date(dateOfBirth);
    }
}

File: Coffers.Public.WebApi.V2/src/Coffers.Public.Queries/Guilds/GamersListView.cs
﻿using System;
using System.Collections.Generic;
using Coffers.Types.Gamer;

namespace Coffers.Public.Queries.Guilds
{
    /// <summary>
    /// base gamers info
    /// </summary>
    public sealed class GamersListView
    {
        /// <summary>
        /// User id
        /// </summary>
        public Guid Id { get; }

        /// <summary>
        /// Characters list
        /// </summary>
        public ICollection<CharacterView> Characters { get; }

        /// <summary>
        /// User balance
        /// </summary>
        public Decimal Balance { get; }

        /// <summary>
        /// Список штрафов
        /// </summary>
        public List<PenaltyView> Penalties { get; }

        /// <summary>
        /// Список займов
        /// </summary>
        public List<LoanView> Loans { get; }

        /// <summary>
        /// Rank
        /// </summary>
        public GamerRank Rank { get; }

        /// <summary>
        /// Статус игрока
        /// </summary>
        public GamerStatus Status { get; }
        /// <summary>
        /// День рождения
        /// </summary>
        public DateTime DateOfBirth { get; }
        /// <summary>
        /// имя юзвера
        /// </summary>
        public String Name { get; }

        public GamersListView(Guid id, String name, Decimal balance, List<CharacterView> characters, GamerRank rank, GamerStatus status, 
            DateTime dateOfBirth, List<PenaltyView> penalties, List<LoanView> loans)
            => (Id, Name, Balance, Characters, Rank, Status, DateOfBirth, Penalties, Loans)
            = (id, name, balance, characters, rank, status, dateOfBirth, penalties, loans);
    }
}